Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnStatus.hs

```diff
@@ -11,42 +11,256 @@
+rideBookingTransaction bookingNewStatus rideNewStatus booking rideEntity = do
+  unless (booking.status == bookingNewStatus) $ do
+    QB.updateStatus booking.id bookingNewStatus
+  case rideEntity of
+    UpdatedRide {ride, rideOldStatus} -> do
+      unless (rideOldStatus == rideNewStatus) $ do
+        QRide.updateMultiple ride.id ride
+    RenewedRide renewedRide -> do
+      QRide.create renewedRide
+
+isStatusChanged :: DB.BookingStatus -> DB.BookingStatus -> RideEntity -> Bool
+isStatusChanged bookingOldStatus bookingNewStatus rideEntity = do
+  let bookingStatusChanged = bookingOldStatus == bookingNewStatus
+  let rideStatusChanged = case rideEntity of
+        UpdatedRide {ride, rideOldStatus} -> rideOldStatus == ride.status
+        RenewedRide {} -> True
+  bookingStatusChanged || rideStatusChanged
+
+onStatus :: (MonadFlow m, CacheFlow m r, EsqDBFlow m r) => DOnStatusReq -> m ()
+onStatus req = do
+  booking <- QB.findByBPPBookingId req.bppBookingId >>= fromMaybeM (BookingDoesNotExist $ "BppBookingId: " <> req.bppBookingId.getId)
+  case req.rideDetails of
+    NewBookingDetails -> do
+      mbExistingRide <- B.runInReplica $ QRide.findActiveByRBId booking.id
+      unless (booking.status == bookingNewStatus) $ do
+        QB.updateStatus booking.id bookingNewStatus
+      whenJust mbExistingRide \existingRide -> do
+        unless (existingRide.status == rideNewStatus) $ do
+          QRide.updateStatus existingRide.id rideNewStatus
+      where
+        bookingNewStatus = DB.NEW
+        rideNewStatus = DRide.CANCELLED
+    RideAssignedDetails {newRideInfo} -> do
+      rideEntity <- buildRideEntity booking updateNewRide newRideInfo
+      rideBookingTransaction bookingNewStatus rideNewStatus booking rideEntity
+      where
+        bookingNewStatus = DB.TRIP_ASSIGNED
+        rideNewStatus = DRide.NEW
+        updateNewRide newRide = newRide{status = rideNewStatus}
+    RideStartedDetails {newRideInfo, rideStartedInfo} -> do
+      rideEntity <- buildRideEntity booking updateRideStarted newRideInfo
+      rideBookingTransaction bookingNewStatus rideNewStatus booking rideEntity
+      where
+        bookingNewStatus = DB.TRIP_ASSIGNED
+        rideNewStatus = DRide.INPROGRESS
+        updateRideStarted newRide =
+          newRide{status = rideNewStatus,
+                  rideStartTime = Just rideStartedInfo.rideStartTime,
+                  driverArrivalTime = rideStartedInfo.driverArrivalTime
+                 }
+    RideCompletedDetails {newRideInfo, rideStartedInfo, rideCompletedInfo} -> do
+      rideEntity <- buildRideEntity booking updateRideCompleted newRideInfo
+      rideBookingTransaction bookingNewStatus rideNewStatus booking rideEntity
+      when (isStatusChanged booking.status bookingNewStatus rideEntity) $ do
+        breakups <- traverse (buildFareBreakup booking.id) rideCompletedInfo.fareBreakups
+        QFareBreakup.deleteAllByBookingId booking.id
+        QFareBreakup.createMany breakups
```

Comment:
Also here we are deleting FareBreakups by booking.id and then creating new ones. Can we just update the existing fare breakups in this case instead of deleting and creating new ones.

Reply:
I was thinking about that. then we need to update each breakup by description (id generated on bap side, did not come from bpp). What if changed number of breakups, or descriptions? logic will be too complicated

Anyway this logic works when status changed, so most likely we will not delete anything, only create

Reply:
Sure @roman-bodavskiy Works